src/cpu/x64/injectors/jit_uni_binary_injector.cpp

```diff
@@ -517,42 +519,82 @@ void jit_uni_binary_injector_t<isa, Vmm>::compute_vector_range(
     static const Vmm zero_vmm(0);
     if (post_op.is_prelu() && is_avx512_) push_opmask(host_, get_aux_kmask());
 
-    Xbyak::Address rhs_arg_addr(0);
+    Xbyak::Address rhs1_arg_addr(0);
+    Xbyak::Address rhs2_arg_addr(0);
 
     // Phase 3 Apply binary post-op over all vmms.
     for (const auto vmm_idx : vmm_idxs) {
         const bool is_start_idx = vmm_idx == start_idx;
-        if (is_start_idx
-                || rhs_arg_params_differ(vmm_idx, vmm_idx - 1, rhs_arg_params,
-                        rhs_broadcasting_strategy)) {
-            rhs_arg_addr = prepare_rhs_arg_addr(vmm_idx, rhs_arg_idx, post_op,
-                    rhs_arg_params, rhs_broadcasting_strategy, is_start_idx);
-        }
-
-        const auto local_vmm_preservation = should_preserve_vmm(
-                vmm_idx, vmm_hint, max_vmm_idx, dt_helper_vmm_needed);
-        const bool &vmm_preservation_needed = local_vmm_preservation.first;
-        const Vmm dst_vmm(vmm_idx);
         const bool with_tail = rhs_arg_static_params_.is_tail
                 && vmm_tail_idx.find(vmm_idx) != vmm_tail_idx.cend()
                 && IMPLICATION(rhs_broadcasting_strategy
                                 == broadcasting_strategy_t::scalar,
                         rhs_arg_static_params_.use_exact_tail_scalar_bcast);
 
+        const Vmm tern_tmp_vmm(rhs_arg_static_params_.rhs_dt_helper_vmm_idx);
+        const auto local_vmm_preservation = should_preserve_vmm(
+                vmm_idx, vmm_hint, max_vmm_idx, dt_helper_vmm_needed);
+        const bool &vmm_preservation_needed = local_vmm_preservation.first;
+        const Vmm dst_vmm(vmm_idx);
+
+        // For binary ops with ternary inputs, a temporary vmm will be needed
+        // for the additional input.
+        if (needs_ternary_input) {
+            const auto rhs2_arg_data_type
+                    = get_src2_desc(post_op, dst_d).data_type;
+            rhs2_arg_addr = prepare_rhs_arg_addr(vmm_idx, rhs_arg_idx + 1,
+                    post_op, rhs_arg_params,
+                    broadcasting_strategy_t::no_broadcast, true, true);
```

Comment:
To pass `no_broadcast`, there must be a check added to `validate_binary_with_dst_consistency`.